Sources/NIOPosix/SocketChannel.swift

```diff
@@ -668,9 +680,44 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
             return true
         }
     }
+
     /// Buffer a write in preparation for a flush.
     override func bufferPendingWrite(data: NIOAny, promise: EventLoopPromise<Void>?) {
-        let data = self.unwrapData(data, as: AddressedEnvelope<ByteBuffer>.self)
+        if let envelope = self.tryUnwrapData(data, as: AddressedEnvelope<ByteBuffer>.self) {
+            return bufferPendingAddressedWrite(data: envelope, promise: promise)
+        }
+        if let data = self.tryUnwrapData(data, as: ByteBuffer.self) {
+            return bufferPendingUnaddressedWrite(data: data, promise: promise)
+        }
+        // TODO: should this be a preconditionFailure or just fail the promise?
+        preconditionFailure("bufferPendingWrite(data:promise:) called with unexpected type.")
+    }
+
+    /// Buffer a write in preparation for a flush.
+    private func bufferPendingUnaddressedWrite(data: ByteBuffer, promise: EventLoopPromise<Void>?) {
+        // It is only appropriate to not use an AddressedEnvelope if the socket is connected.
+        guard self.remoteAddress != nil else {
+            promise?.fail(IOError(errnoCode: ENOTCONN, reason: """
+                Address not supplied when writing to an unconnected socket
+                """))
+            return
+        }
+
+        if !self.pendingWrites.add(data: data, promise: promise) {
+            assert(self.isActive)
+            self.pipeline.syncOperations.fireChannelWritabilityChanged()
+        }
+    }
+
+    /// Buffer a write in preparation for a flush.
+    private func bufferPendingAddressedWrite(data: AddressedEnvelope<ByteBuffer>, promise: EventLoopPromise<Void>?) {
+        // It is only appropriate to use an AddressedEnvelope if the socket is not connected.
+        guard self.remoteAddress == nil else {
+            promise?.fail(IOError(errnoCode: EISCONN, reason: """
```

Comment:
Here we should only error if the remote address doesn't match the address in the envelope, but otherwise it's totally fine.

Reply:
Resolved by https://github.com/apple/swift-nio/pull/2084/commits/eb580d69ca04991d91b7324fee725ca25bce9af9.